File: packages/core/src/thread.ts
import { z } from "zod";

/**
 * Thread identifier for per-workflow state and checkpointing.
 */
export const ThreadIdSchema = z.string().min(1);
export type ThreadId = z.infer<typeof ThreadIdSchema>;

/**
 * Snapshot of graph state at a node (for checkpointing).
 */
export interface ThreadStateSnapshot {
  threadId: string;
  nodeId: string;
  state: Record<string, unknown>;
  timestamp: number;
  version?: number;
}

File: python/src/agent_framework/thread.py
"""Thread and state snapshot types."""

from typing import Any


ThreadId = str


class ThreadStateSnapshot:
    """Snapshot of graph state at a node for checkpointing."""

    def __init__(
        self,
        thread_id: str,
        node_id: str,
        state: dict[str, Any],
        timestamp: float,
        version: int | None = None,
    ):
        self.thread_id = thread_id
        self.node_id = node_id
        self.state = state
        self.timestamp = timestamp
        self.version = version
